### PR TITLE
pass info

### PR DESCRIPTION
**Pull request by olliethomas**
Basically, I want to be able to pass additional information from an Async Algo to the main calling script. 
There's already an `info` parameter, so now just write this value to json, then load if exists.

### CODE REVIEW

File: tests/saving_data_test.py [hunk 1]
```diff
@@ -1,20 +1,16 @@
-"""
-Test the saving data capability
-"""
+"""Test the saving data capability."""
```
Comment (olliethomas) [inline tests/saving_data_test.py:1]: There's a few of these. I have a pydocstyle pre-commit hook, so have to add this kind of thing.

File: tests/saving_data_test.py [hunk 2]
```diff
@@ -48,10 +44,32 @@ def _script_command(self, train_path, _, pred_path):
     pd.testing.assert_series_equal(data_tuple.x["a1"], data_x.hard, check_names=False)
 
 
+def test_preidtions_loaded(temp_dir) -> None:
```
Comment (olliethomas) [inline tests/saving_data_test.py:47]:
```suggestion
def test_predictions_loaded(temp_dir) -> None:
```

File: pyproject.toml [hunk 3]
```diff
@@ -43,27 +43,6 @@ skip-string-normalization = true
...
-  "setuptools",
-  "sklearn",
-  "tap",
-  "teext",
-  "torch",
-  "torchvision",
-  "tqdm",
-  "typing_extensions",
-]
+known_third_party = ["PIL", "black", "fairlearn", "git", "matplotlib", "mypy", "numpy", "pandas", "pylint", "pytest", "scipy", "seaborn", "setuptools", "sklearn", "tap", "teext", "torch", "torchvision", "tqdm", "typing_extensions"]
```
Comment (olliethomas) [inline pyproject.toml:46]: my isort plugin only started doing this in the pyproject.toml. It didn't do it in isort.config.
Maybe there's a setting I have that's not right

**Comment by thomkeh** [comment]
Hm, but I thought I configured pydocstyle to not require docstrings in the `__init__` (requiring it only in the class itself)... maybe I didn't configure it in the right place.

File: ethicml/utility/data_structures.py [hunk 4]
```diff
@@ -230,21 +234,33 @@ def info(self) -> Dict[str, float]:
...
-            return Prediction(hard=pd.Series(np.squeeze(data["hard"])))
+                return SoftPrediction(soft=pd.Series(np.squeeze(data["soft"])), info=info)
+            return Prediction(hard=pd.Series(np.squeeze(data["hard"])), info=info)
 
     def to_npz(self, npz_path: Path) -> None:
         """Save prediction as npz file."""
+        if self.info:
+            for v in self.info.values():
+                assert isinstance(v, float), "Info must be Dict[str, float]"
+            json_path = npz_path.parent / 'info.txt'
```
Comment (thomkeh) [inline ethicml/utility/data_structures.py:252]: `.json` maybe? not that important though

```suggestion
            json_path = npz_path.parent / 'info.json'
```

**Comment by olliethomas** [comment]
> Hm, but I thought I configured pydocstyle to not require docstrings in the `__init__` (requiring it only in the class itself)... maybe I didn't configure it in the right place.

or.... I have not made my pre-commit hook have the same requirements

**Review by thomkeh — APPROVED** [review]
I think it would have been possible to store the dictionary in the npz file together with the predictions, but it's fine.

**Comment by olliethomas** [comment]
> I think it would have been possible to store the dictionary in the npz file together with the predictions, but it's fine.

this would be nicer - adding a .json file to the parent feels pretty bad

**Pull request merged**